atlasdb-coordination-impl/src/main/java/com/palantir/atlasdb/coordination/CoordinationStore.java

```diff
@@ -0,0 +1,64 @@
+/*
+ * (c) Copyright 2018 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.coordination;
+
+import java.util.Optional;
+
+import com.palantir.atlasdb.keyvalue.impl.CheckAndSetResult;
+
+import okio.ByteString;
+
+/**
+ * A {@link CoordinationStore} stores data that a {@link CoordinationService} may use.
+ */
+public interface CoordinationStore {
+    /**
+     * Gets the value associated with a given sequence number.
+     * Sequence numbers are expected to be strictly positive.
+     *
+     * @param sequenceNumber sequence number to read a value for
+     * @return value stored at that sequence number; empty if none is present
+     */
+    Optional<ByteString> getValue(long sequenceNumber);
+
+    /**
+     * Stores a value for a given sequence number.
+     * Sequence numbers are expected to be strictly positive.
+     *
+     * @param sequenceNumber sequence number to store a value for
+     * @param value value to be stored
+     */
+    void putValue(long sequenceNumber, ByteString value);
```

Comment:
We should be clear about repeated puts for the same sequenceNumber with different values not being supported. Probably fine to say it's nondeterministic bit idempotent and leave it at that?

Reply:
Does it need to be nondeterministic? Can we just fail instead?

edit: I see that we can't, since the table is non-transactional and the values are potentially too big to CAS. Might still be worth best effort protection, since a race is pretty unlikely? Also unclear to me that not failing is acceptable, since this could cause corruption.

Reply:
@sandorw I think I will change this to failing. It _is_ possible to fail, actually (to be clear: the issue blocks CAS, but _not_ PUE as far as I know - I will confirm this with a test).

Thinking about this, values shouldn't be written too often (the bound is written a lot, but you don't write a new value unless you must), so it's probably fair to incur the cost of one round of Cassandra's Paxos when doing this